File: soungmunkim/C++/BinaryTree_BST_RBT/199_RightSideViewBinarytree.cpp
/*
[binary tree의 오른쪽 side node만 프린트하기]

Given the root of a binary tree, imagine yourself standing on the right side of it, 
return the values of the nodes you can see ordered from top to bottom.

(base) soungmunkim@gimseongmun-ui-MacBookPro BinaryTree_BST_RBT % g++ -std=c++17 199_RightSideViewBinarytree.cpp -o test
(base) soungmunkim@gimseongmun-ui-MacBookPro BinaryTree_BST_RBT % ./test
Enter node values in [a,b,c,...] format: [1,2,3,None,5,None,4]
[1, 3, 4]

(base) soungmunkim@gimseongmun-ui-MacBookPro BinaryTree_BST_RBT % ./test
Enter node values in [a,b,c,...] format: [1,None,3]
[1, 3]

(base) soungmunkim@gimseongmun-ui-MacBookPro BinaryTree_BST_RBT % ./test
Enter node values in [a,b,c,...] format: []
[]
*/

/*------------- 사용자로부터 리스트 입력받고 사용할때 (None으로 입력 포함) ------------------*/
#include <iostream>
#include <vector>
#include <queue>
#include <sstream>
#include <string>

using namespace std;

// TreeNode 클래스 정의
class TreeNode {
public:
    int val;                // 노드의 값을 저장하는 변수입니다.
    TreeNode* left;         // 왼쪽 자식 노드를 가리키는 포인터입니다.
    TreeNode* right;        // 오른쪽 자식 노드를 가리키는 포인터입니다.
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}  // 생성자 함수. 초기값 x를 받아서 노드의 값을 설정하고, 자식 노드들은 NULL로 초기화합니다.
};

// 주어진 벡터를 이진 트리로 변환하는 함수
TreeNode* vectorToTree(vector<int>& nums) {
    if (nums.empty()) return nullptr;  // 빈 벡터이면 루트 노드는 NULL입니다.

    queue<TreeNode*> nodes;  // TreeNode 포인터 큐를 생성합니다.
    TreeNode* root = new TreeNode(nums[0]);  // 루트 노드를 생성합니다.
    nodes.push(root);  // 루트 노드를 큐에 추가합니다.

    int index = 1;  // 벡터 인덱스를 초기화합니다.
    while (!nodes.empty() && index < nums.size()) {
        TreeNode* current = nodes.front();  // 현재 노드를 큐에서 꺼냅니다.
        nodes.pop();  // 현재 노드를 큐에서 제거합니다.

        // 왼쪽 자식 노드 생성 및 연결
        if (nums[index] != -1) {
            current->left = new TreeNode(nums[index]);
            nodes.push(current->left);
        }
        index++;

        // 오른쪽 자식 노드 생성 및 연결
        if (index < nums.size() && nums[index] != -1) {
            current->right = new TreeNode(nums[index]);
            nodes.push(current->right);
        }
        index++;
    }

    return root;  // 생성된 이진 트리의 루트 노드를 반환합니다.
}

// 이진 트리에서 오른쪽 뷰 노드 값 추출 함수
vector<int> rightSideView(TreeNode* root) {
    if (!root) return vector<int>();

    queue<TreeNode*> q;
    q.push(root);
    vector<int> ans;

    while (!q.empty()) {
        int levelSize = q.size();

        for (int i = 0; i < levelSize; i++) {
            TreeNode* a = q.front();
            q.pop();

            if (a->left) q.push(a->left);
            if (a->right) q.push(a->right);

            if (i == levelSize - 1) {
                ans.push_back(a->val);
            }
        }
    }

    return ans;
}

int main() {
     // 사용자에게 숫자 리스트를 입력하라는 메시지를 출력
    cout << "Enter node values in [a,b,c,...] format: ";
    vector<int> input; // 입력을 저장할 벡터
    string line;
    getline(cin, line); // 입력 받기

    // 괄호와 쉼표를 제거하고 파싱
    // 첫 번째와 마지막 문자를 제외한 나머지 부분의 문자열
    stringstream ss(line.substr(1, line.size() - 2)); 
    string item;


    while (getline(ss, item, ',')) {
        // 문자열이 "None"이 아니면 리스트에 추가
        if (item != "None") {
            input.push_back(stoi(item));
        }
    }
    // 연결 리스트 생성 및 출력
    TreeNode* head = vectorToTree(input);

    vector<int> result = rightSideView(head);

    // 최종 결과 프린트하기
    cout << "[";
    for (int i = 0; i < result.size(); i++){
        cout  << result[i];
        if(i != result.size() - 1){
            cout << ", ";
        }
    }
    cout << "]" << endl;

    return 0;
}

File: soungmunkim/Python/BinaryTree/199_RightSideViewBinarytree.py
"""
[binary tree의 오른쪽 side node만 프린트하기]

Given the root of a binary tree, imagine yourself standing on the right side of it, 
return the values of the nodes you can see ordered from top to bottom.

root = [1,2,3,None,5,None,4]
tree = list_to_btree(root)
rightSideView(tree)
[1, 3, 4]

root = [1,None,3]
tree1 = list_to_btree(root)
rightSideView(tree1)
[1, 3

root = []
tree2 = list_to_btree(root)
rightSideView(tree2)]
[]
"""

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 숫자 리스트를 이진 트리로 변환하는 함수
def list_to_btree(nums: list) -> TreeNode:
    if not nums:
        return None

    # 노드들을 저장할 큐를 생성합니다.
    queue = []
    
    # 루트 노드를 생성하고 큐에 추가합니다.
    root = TreeNode(nums[0])
    queue.append(root)
    
    # 노드들을 연결하기 위해 리스트를 순회합니다.
    i = 1
    while queue and i < len(nums):
        # 현재 노드를 큐에서 꺼냅니다.
        current = queue.pop(0)
        
        # 왼쪽 자식이 있다면, 연결하고 큐에 추가합니다.
        if nums[i] is not None:
            left_child = TreeNode(nums[i])
            current.left = left_child
            queue.append(left_child)
        i += 1

        # 오른쪽 자식이 있다면, 연결하고 큐에 추가합니다.
        if i < len(nums) and nums[i] is not None:
            right_child = TreeNode(nums[i])
            current.right = right_child
            queue.append(right_child)
        i += 1

    return root

def rightSideView(root:TreeNode):
    if not root:
        return []
    
    q=[root] #root는 treenode이기 대문에 root노드 하나만 들어감 (초기화)
    ans =[]
    # queue가 빌때까지 둘기
    while q:
        for i in range(len(q)): #bfs로 돌기 right는 맨 마지막에 넣고 left부터 꺼내는 방식
            a=q.pop(0)
            if a.left:
                q.append(a.left)
            if a.right:
                q.append(a.right) 
        # 왼쪽 먼저 넣고 왼쪽 먼저 빼는 방식이니까 맨 마지막 빼는 건 항상 right side 쪽
        ans.append(a.val)
    return ans

if __name__ == "__main__":
    root = [1,2,3,None,5,None,4]
    tree = list_to_btree(root)
    result = rightSideView(tree)
    print(result)  # 결과 출력